nixos/modules/services/networking/headscale.nix

```diff
@@ -0,0 +1,462 @@
+      domains = cfg.dns.domains;
+      magic_dns = cfg.dns.magicDns;
+      base_domain = cfg.dns.baseDomain;
+    };
+
+    unix_socket = "${runDir}/headscale.sock";
+
+    # OpenID Connect
+    oidc = {
+      issuer = cfg.openIdConnect.issuer;
+      client_id = cfg.openIdConnect.clientId;
+      domain_map = cfg.openIdConnect.domainMap;
+    };
+
+    tls_letsencrypt_cache_dir = "${dataDir}/.cache";
+
+    acl_policy_path = cfg.aclPolicyFile;
+  } // optionalAttrs (cfg.database.host != null) {
+    db_host = cfg.database.host;
+  } // optionalAttrs (cfg.database.port != null) {
+    db_port = cfg.database.port;
+  } // optionalAttrs (cfg.database.name != null) {
+    db_name = cfg.database.name;
+  } // optionalAttrs (cfg.database.user != null) {
+    db_user = cfg.database.user;
+  } // optionalAttrs (cfg.tls.letsencrypt.hostname != null) {
+    tls_letsencrypt_hostname = cfg.tls.letsencrypt.hostname;
+  } // optionalAttrs (cfg.tls.letsencrypt.challengeType != null) {
+    tls_letsencrypt_challenge_type = cfg.tls.letsencrypt.challengeType;
+  } // optionalAttrs (cfg.tls.letsencrypt.listen != null) {
+    tls_letsencrypt_listen = cfg.tls.letsencrypt.listen;
+  } // optionalAttrs (cfg.tls.certFile != null) {
+    tls_cert_path = cfg.tls.certFile;
+  } // optionalAttrs (cfg.tls.keyFile != null) {
+    tls_key_path = cfg.tls.keyFile;
+  };
+
+  settingsFormat = pkgs.formats.yaml { };
+  configFile = settingsFormat.generate "headscale.yaml" settings;
+in
+{
+  options = {
+    services.headscale = {
+      enable = mkEnableOption "headscale, Open Source coordination server for Tailscale";
+
+      package = mkPackageOption pkgs "headscale";
+
+      user = mkOption {
+        default = "headscale";
+        type = types.str;
+        description = ''
+          User the headscale server should execute under.
+        '';
+      };
+
+      group = mkOption {
+        default = "headscale";
+        type = types.str;
+        description = ''
+          If the default user "headscale" is configured then this is the primary
+          group of that user.
+        '';
+      };
+
+      serverUrl = mkOption {
+        type = types.str;
+        default = "http://127.0.0.1:8080";
+        description = ''
+          The url clients will connect to.
+        '';
+        example = "https://myheadscale.example.com:443";
+      };
+
+      address = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Listening address of headscale
+        '';
+        example = "0.0.0.0";
+      };
+
+      port = mkOption {
+        type = types.int;
+        default = 8080;
+        description = ''
+          Listening port of headscale
+        '';
+        example = "443";
+      };
+
+      privateKeyFile = mkOption {
+        type = types.path;
+        description = ''
+          Path to private key file, generated automatically if it does not exist.
+        '';
+      };
```

Comment:
I am actually contemplating if there should be a:

```
default = "${dataDir}/private.key";
```

My reasoning is that if you dont have a key, headscale will create one. ATM there is no headscale cli/tool to create this key to put it in your secret management.

I have filed a ticket for such a CLI option to exist: https://github.com/juanfont/headscale/issues/286

Reply:
sounds like a good idea if having the key in secret management is not crucial (eg because it has to stable over long periods).

Reply:
Yes, it is needed, but it can be rotated "at will" if a user desire. So it will be automatically generated if it doesnt exist.